src/metabase/query_processor.clj

```diff
@@ -66,7 +66,13 @@
   Processor methods. This function takes the fully pre-processed query, runs it, and returns the results, which then
   run through the various post-processing steps."
   [query :- (s/pred map?)]
-  (driver/execute-query driver/*driver* query))
+  (if-let [f (:data-fn query)]
+    (let [query (dissoc query :data-fn)]
+      (driver/execute-query-callback
+        driver/*driver*
+        query
+        (fn [res] (f ((format-rows/format-rows (annotate/add-column-info (constantly res))) query)))))
```

Comment:
It's a bit odd to me that the decoration of the results with the column info and row formatting, be done here. Would it not be more appropriate to do this higher up the call stack? Maybe it should even be defined as part of `f`, so that `execute-query` doesn't have to know about formatting. What do you think?

Reply:
This is the one wart I couldn't figure out how to get rid of. I think ideally, `f` would receive the same fully decorated result that one would get when no callback is used at all. This poses 2 problems:

1. I don't know how to do this, short of fully rewriting the middleware stack.
2. At least one middleware - the one that adds the row count - would result in the entire result set being loaded into memory, defeating the purpose.

I settled on making ad hoc calls to the middleware necessary for export. The alternative, which I believe you are alluding to, would be to give up and hand the raw results directly to `f`, which could then invoke the middleware.  Calling into query processor middleware internals from export code felt more wrong to me than providing an incomplete set of middleware to all consumers. I can certainly flip it around if requested, but I hope there's a third option I'm not seeing.

Reply:
> Calling into query processor middleware internals from export code felt more wrong to me...

Agreed! Maybe this calls for a different pipeline to be defined, for using in async/export cases like these?

I think I'll defer to @camsaul's knowledge of the intricacies of the query processor.